### PR TITLE
paramiko.Channel - changed type of environment variables to bytes

### PR DESCRIPTION
**Pull request by manfred-kaiser**

### CODE REVIEW

**Comment by manfred-kaiser** [comment]
There is one error, but I have not changed anything related to it. Perhaps the error comes from the upstram repo?

```
Run ./tests/pytype_test.py
Traceback (most recent call last):
  File "/home/runner/work/typeshed/typeshed/./tests/pytype_test.py", line 20, in <module>
    from pytype import config as pytype_config, load_pytd
  File "/opt/hostedtoolcache/Python/3.9.8/x64/lib/python3.9/site-packages/pytype/config.py", line 14, in <module>
    from pytype import errors
  File "/opt/hostedtoolcache/Python/3.9.8/x64/lib/python3.9/site-packages/pytype/errors.py", line 13, in <module>
    from pytype import matcher
  File "/opt/hostedtoolcache/Python/3.9.8/x64/lib/python3.9/site-packages/pytype/matcher.py", line 14, in <module>
    from pytype.overlays import typing_overlay
  File "/opt/hostedtoolcache/Python/3.9.8/x64/lib/python3.9/site-packages/pytype/overlays/typing_overlay.py", line 13, in <module>
    from pytype.overlays import collections_overlay
  File "/opt/hostedtoolcache/Python/3.9.8/x64/lib/python3.9/site-packages/pytype/overlays/collections_overlay.py", line 11, in <module>
    from pytype.pyi import parser
  File "/opt/hostedtoolcache/Python/3.9.8/x64/lib/python3.9/site-packages/pytype/pyi/parser.py", line 11, in <module>
    from pytype.pyi import classdef
  File "/opt/hostedtoolcache/Python/3.9.8/x64/lib/python3.9/site-packages/pytype/pyi/classdef.py", line 7, in <module>
    from pytype.pyi.types import ParseError  # pylint: disable=g-importing-member
  File "/opt/hostedtoolcache/Python/3.9.8/x64/lib/python3.9/site-packages/pytype/pyi/types.py", line 11, in <module>
    from typed_ast import ast3
  File "/opt/hostedtoolcache/Python/3.9.8/x64/lib/python3.9/site-packages/typed_ast/ast3.py", line 40, in <module>
    from typed_ast import _ast3
ImportError: /opt/hostedtoolcache/Python/3.9.8/x64/lib/python3.9/site-packages/typed_ast/_ast3.cpython-39-x86_64-linux-gnu.so: undefined symbol: _PyUnicode_DecodeUnicodeEscape
Error: Process completed with exit code 1.
```

**Comment by srittau** [comment]
Waiting for new typed-ast version.

**Comment by srittau** [comment]
This will need a merge of `master`.

**Comment by manfred-kaiser** [comment]
> This will need a merge of `master`.

Thanks!

There are some type hints, which might be wrong. I will check them if possible and convert this draft to a pull request.

My application does not use all methods, so it's not easy to get the reals type of an argument.

File: stubs/paramiko/paramiko/channel.pyi [hunk 1]
```diff
@@ -41,14 +41,14 @@ class Channel(ClosingContextManager):
     def __init__(self, chanid: int) -> None: ...
     def __del__(self) -> None: ...
     def get_pty(
-        self, term: str = ..., width: int = ..., height: int = ..., width_pixels: int = ..., height_pixels: int = ...
+        self, term: bytes = ..., width: int = ..., height: int = ..., width_pixels: int = ..., height_pixels: int = ...
```
Comment (srittau) [inline stubs/paramiko/paramiko/channel.pyi:44]: The default is a `str`:

https://github.com/paramiko/paramiko/blob/a598ca14d6451343c69be97e0e1c65eb15c0e607/paramiko/channel.py#L166

As this is passed to `add_string()`, which accepts both `str` and `bytes`, I suggest to use:

```suggestion
        self, term: str | bytes = ..., width: int = ..., height: int = ..., width_pixels: int = ..., height_pixels: int = ...
```

File: stubs/paramiko/paramiko/channel.pyi [hunk 2]
```diff
@@ -41,14 +41,14 @@ class Channel(ClosingContextManager):
...
     ) -> None: ...
     def invoke_shell(self) -> None: ...
     def exec_command(self, command: str) -> None: ...
     def invoke_subsystem(self, subsystem: str) -> None: ...
     def resize_pty(self, width: int = ..., height: int = ..., width_pixels: int = ..., height_pixels: int = ...) -> None: ...
-    def update_environment(self, environment: Mapping[str, str]) -> None: ...
-    def set_environment_variable(self, name: str, value: str) -> None: ...
+    def update_environment(self, environment: Mapping[bytes, bytes]) -> None: ...
+    def set_environment_variable(self, name: bytes, value: bytes) -> None: ...
```
Comment (srittau) [inline stubs/paramiko/paramiko/channel.pyi:51]: Same here for both `name` and `value`.

Comment (manfred-kaiser) [reply stubs/paramiko/paramiko/channel.pyi:51]: Thanks for the information.

The arguments of ``paramiko.ServerInterface.check_channel_env_request`` are `bytes`. 

In my application the arguments are typechecked with typeguard and the types should be correct.

https://github.com/ssh-mitm/ssh-mitm/blob/8bffa3b8e309887b8729a9fede7f233f7c85f4e2/ssh_proxy_server/interfaces/server.py#L265:L266

The received arguments are passed to `paramiko.Channel.set_environment_variable`. So there is a difference in the argument types of both methods.

Perhaps the this was my mistake and I assumed, that both methods accept the same types.

I think this needs further tests. I will check that.

Comment (manfred-kaiser) [reply stubs/paramiko/paramiko/channel.pyi:51]: `paramiko.Channel.set_environment_variable` creates a new message and adds the name and the vallue with the method add_string.

https://github.com/paramiko/paramiko/blob/a598ca14d6451343c69be97e0e1c65eb15c0e607/paramiko/channel.py#L360:L361

The method add_string converts the string to bytes: https://github.com/paramiko/paramiko/blob/a598ca14d6451343c69be97e0e1c65eb15c0e607/paramiko/message.py#L273

To convert the value to bytes, the method asbytes is used: https://github.com/paramiko/paramiko/blob/a598ca14d6451343c69be97e0e1c65eb15c0e607/paramiko/common.py#L193

asbytes uses the function `b`: https://github.com/paramiko/paramiko/blob/a598ca14d6451343c69be97e0e1c65eb15c0e607/paramiko/py3compat.py#L135

The function `b` is able to use `str` and `bytes`.

The same is true for `get_pty`.

**So a lot of types have to be changed to `str | bytes`** :disappointed_relieved:

Comment (srittau) [reply stubs/paramiko/paramiko/channel.pyi:51]: Don't feel obligated to change anything that you don't need. Any gradual improvement is welcome!

Comment (manfred-kaiser) [reply stubs/paramiko/paramiko/channel.pyi:51]: :+1:

**Comment by manfred-kaiser** [comment]
I have updated some typehints. The changes are done by reviewing the code. Most of the methods are using Message.add_string, which is able to handle bytes and str.

At the moment only methods where changed, which I'm using in my application and where I'm sure, that the types are correct.

**Review by srittau — APPROVED** [review]
Thanks!

**Pull request merged**